File: reference/provenance-cs/src/Receipt.cs
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;

namespace Raucle.Provenance;

/// <summary>
/// Raucle Provenance Receipt v1 — C# reference implementation.
/// Spec: https://raucle.com/spec/provenance/v1
///
/// Mirrors the canonical Python reference (raucle_detect/provenance.py)
/// byte-for-byte: same JOSE header (incl. the "raucle/v1":"provenance"
/// tag), same payload field set, string-typed model/tool/corpus,
/// sha256:-prefixed hashes, and the same content-addressed id
/// ("sha256:" + hex(sha256(jws))). A receipt emitted here verifies in
/// the other reference implementations and yields the identical id.
///
/// Ed25519 comes from BouncyCastle (the .NET BCL has no Ed25519);
/// SHA-256 + base64url from the BCL.
/// </summary>
public static class Receipt
{
    private const string JwsTyp = "provenance-receipt/v1";
    private const string Iss = "raucle-detect/provenance";

    private static readonly HashSet<string> ValidOperations = new()
    {
        "user_input", "model_call", "tool_call", "retrieval",
        "guardrail_scan", "agent_handoff", "sanitisation", "merge",
    };
    private static readonly HashSet<string> KnownFields = new()
    {
        "iss", "typ", "iat", "agent_id", "agent_key_id", "operation", "parents",
        "input_hash", "output_hash", "taint", "ruleset_hash",
        "guardrail_verdict", "model", "tool", "corpus", "tenant",
    };

    // ── base64url ──────────────────────────────────────────────────

    public static string B64UrlEncode(byte[] b) =>
        Convert.ToBase64String(b).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    public static byte[] B64UrlDecode(string s)
    {
        var t = s.Replace('-', '+').Replace('_', '/');
        switch (t.Length % 4) { case 2: t += "=="; break; case 3: t += "="; break; }
        return Convert.FromBase64String(t);
    }

    public static string Sha256Hex(byte[] b) =>
        Convert.ToHexString(SHA256.HashData(b)).ToLowerInvariant();

    // ── validation (§4) ────────────────────────────────────────────
    // Lenient where the Python reference is lenient: enforces structural
    // invariants (typ literal, required fields per operation, parent
    // rules), not value shapes.

    public static void Validate(JObj p)
    {
        foreach (var k in p.Members.Keys)
            if (!KnownFields.Contains(k) && !k.StartsWith("x_"))
                throw new ProvException($"reserved/unknown field: {k} (§14)");

        if (p.Str("typ") != JwsTyp)
            throw new ProvException($"payload typ must be {JwsTyp}");

        var op = p.Str("operation") ?? "";
        if (!ValidOperations.Contains(op)) throw new ProvException($"unknown operation: {op}");

        bool Has(string key) => !string.IsNullOrEmpty(p.Str(key));

        if (op == "guardrail_scan" && !Has("guardrail_verdict"))
            throw new ProvException("guardrail_scan requires guardrail_verdict (§4)");
        if (op == "guardrail_scan" && !Has("ruleset_hash"))
            throw new ProvException("guardrail_scan requires ruleset_hash (§4)");
        if (op == "model_call" && !Has("model"))
            throw new ProvException("model_call requires model (§4)");
        if ((op is "tool_call" or "sanitisation") && !Has("tool"))
            throw new ProvException($"{op} requires tool (§4)");
        if ((op is "retrieval" or "sanitisation") && !Has("corpus"))
            throw new ProvException($"{op} requires corpus (§4)");

        var parents = p.StrArray("parents") ?? new List<string>();
        if (op == "user_input" && parents.Count > 0) throw new ProvException("user_input must have no parents");
        if (op != "user_input" && parents.Count == 0) throw new ProvException($"{op} requires at least one parent");

        // §4.2/§4.3.1: parents and taint MUST be sorted in UTF-16 code-unit
        // order and unique. StringComparer.Ordinal is UTF-16; a strictly-
        // increasing check enforces sorted AND unique at once.
        foreach (var name in new[] { "parents", "taint" })
        {
            var arr = p.StrArray(name);
            if (arr == null) continue;
            for (int i = 1; i < arr.Count; i++)
                if (StringComparer.Ordinal.Compare(arr[i - 1], arr[i]) >= 0)
                    throw new ProvException(
                        $"{name} must be sorted in UTF-16 code-unit order and unique (§4.3.1)");
        }
    }

    // ── emit / verify ──────────────────────────────────────────────

    public static SignedReceipt Emit(JObj payload, Ed25519PrivateKeyParameters privateKey)
    {
        // Inject the constant iss/typ and sort parents+taint, exactly as
        // Python's ProvenanceReceipt.payload() does.
        payload.Set("iss", JVal.Of(Iss));
        payload.Set("typ", JVal.Of(JwsTyp));
        SortStringArray(payload, "parents");
        SortStringArray(payload, "taint");

        Validate(payload);
        var kid = payload.Str("agent_key_id") ?? throw new ProvException("missing agent_key_id");

        var header = new JObj()
            .Set("alg", JVal.Of("EdDSA"))
            .Set("typ", JVal.Of(JwsTyp))
            .Set("kid", JVal.Of(kid))
            .Set("crit", JVal.Arr(new[] { "raucle/v1" }))
            .Set("raucle/v1", JVal.Of("provenance"));

        var headerB = Canonical.Encode(header);
        var payloadB = Canonical.Encode(payload);
        var signingInput = B64UrlEncode(headerB) + "." + B64UrlEncode(payloadB);

        var signer = new Ed25519Signer();
        signer.Init(true, privateKey);
        var msg = Encoding.ASCII.GetBytes(signingInput);
        signer.BlockUpdate(msg, 0, msg.Length);
        var sig = signer.GenerateSignature();

        var jws = signingInput + "." + B64UrlEncode(sig);
        return new SignedReceipt(jws, payload, "sha256:" + Sha256Hex(Encoding.ASCII.GetBytes(jws)));
    }

    public static SignedReceipt Verify(string jws, Ed25519PublicKeyParameters publicKey)
    {
        var parts = jws.Split('.');
        if (parts.Length != 3) throw new ProvException("JWS must have three segments");
        var (headerB, payloadB, sigB) = (parts[0], parts[1], parts[2]);

        using var headerDoc = JsonDocument.Parse(B64UrlDecode(headerB));
        var header = headerDoc.RootElement;
        if (GetStr(header, "alg") != "EdDSA") throw new ProvException("unsupported alg");
        if (GetStr(header, "typ") != JwsTyp) throw new ProvException("unexpected typ");
        var critOk = header.TryGetProperty("crit", out var crit)
                     && crit.ValueKind == JsonValueKind.Array
                     && crit.GetArrayLength() == 1
                     && crit[0].ValueKind == JsonValueKind.String && crit[0].GetString() == "raucle/v1";
        if (!critOk) throw new ProvException("crit must be exactly ['raucle/v1']");
        if (GetStr(header, "raucle/v1") != "provenance")
            throw new ProvException("header 'raucle/v1' must be 'provenance'");
        var allowedHeaderKeys = new HashSet<string> { "alg", "typ", "kid", "crit", "raucle/v1" };
        foreach (var prop in header.EnumerateObject())
            if (!allowedHeaderKeys.Contains(prop.Name))
                throw new ProvException($"unexpected JOSE header key: {prop.Name}");

        // Canonical byte-equality (spec v1 §4.3, matches the Python reference):
        // the signature binds the on-wire bytes, but a non-canonical header
        // (unsorted keys / extra whitespace) would still verify without
        // re-encoding and comparing — admitting byte-different receipts for the
        // same logical content.
        var canonHeader = Canonical.Encode(JVal.FromJsonElement(header));
        if (!canonHeader.SequenceEqual(B64UrlDecode(headerB)))
            throw new ProvException("JOSE header is not canonical JSON (JCS)");

        var signingInput = Encoding.ASCII.GetBytes(headerB + "." + payloadB);
        var verifier = new Ed25519Signer();
        verifier.Init(false, publicKey);
        verifier.BlockUpdate(signingInput, 0, signingInput.Length);
        if (!verifier.VerifySignature(B64UrlDecode(sigB))) throw new ProvException("signature invalid");

        var payloadBytes = B64UrlDecode(payloadB);
        using var payloadDoc = JsonDocument.Parse(payloadBytes);
        var payload = (JObj)JVal.FromJsonElement(payloadDoc.RootElement);
        if (!Canonical.Encode(payload).SequenceEqual(payloadBytes))
            throw new ProvException("JWS payload is not canonical JSON (JCS)");
        Validate(payload);

        if (GetStr(header, "kid") != payload.Str("agent_key_id"))
            throw new ProvException("header.kid != payload.agent_key_id (§3)");

        return new SignedReceipt(jws, payload, "sha256:" + Sha256Hex(Encoding.ASCII.GetBytes(jws)));
    }

    private static void SortStringArray(JObj p, string key)
    {
        var arr = p.StrArray(key);
        if (arr == null) return;
        var sorted = arr.OrderBy(x => x, StringComparer.Ordinal);
        p.Set(key, JVal.Arr(sorted));
    }

    private static string? GetStr(JsonElement e, string key) =>
        e.TryGetProperty(key, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;
}

/// <summary>A signed receipt: the Compact JWS, the parsed payload, and the content-addressed id.</summary>
public sealed record SignedReceipt(string Jws, JObj Payload, string Id);

public sealed class ProvException : Exception
{
    public ProvException(string message) : base(message) { }
}

File: reference/provenance-ts/src/receipt.ts
/**
 * Raucle Provenance Receipt v1 — TypeScript reference implementation.
 *
 * Spec: https://raucle.com/spec/provenance/v1
 *
 * Mirrors the canonical Python reference (raucle_detect/provenance.py)
 * byte-for-byte: same JOSE header (incl. the `"raucle/v1": "provenance"`
 * tag), same payload field set, string-typed model/tool/corpus,
 * sha256:-prefixed hashes, and the same content-addressed id
 * (`"sha256:" + hex(sha256(jws))`). A receipt emitted here verifies in
 * the other reference implementations and yields the identical id.
 *
 * Zero runtime dependencies — uses Node's webcrypto for Ed25519 and
 * SHA-256.
 */

import { webcrypto } from 'node:crypto'
import { byCodeUnit, canonicalEncode } from './canonical.js'

const subtle = webcrypto.subtle

export const VALID_OPERATIONS = new Set([
  'user_input',
  'model_call',
  'tool_call',
  'retrieval',
  'guardrail_scan',
  'agent_handoff',
  'sanitisation',
  'merge',
])

const JWS_TYP = 'provenance-receipt/v1'
const ISS = 'raucle-detect/provenance'
const JWS_CRIT = ['raucle/v1']

export interface ReceiptPayload {
  iat: number
  agent_id: string
  agent_key_id: string
  operation: string
  parents: string[]
  taint: string[]
  input_hash?: string
  output_hash?: string
  model?: string
  tool?: string
  corpus?: string
  ruleset_hash?: string
  guardrail_verdict?: string
  tenant?: string
  /** Injected by emit; present on parsed payloads. */
  iss?: string
  typ?: string
  /** Extension fields; keys MUST be prefixed `x_` (spec §14). */
  [extra: string]: unknown
}

const KNOWN_FIELDS = new Set([
  'iss', 'typ', 'iat', 'agent_id', 'agent_key_id', 'operation', 'parents',
  'taint', 'input_hash', 'output_hash', 'model', 'tool', 'corpus',
  'ruleset_hash', 'guardrail_verdict', 'tenant',
])

export interface Receipt {
  /** The Compact JWS string. */
  jws: string
  /** Parsed payload. */
  payload: ReceiptPayload
  /** Content-addressed id: "sha256:" + hex SHA-256 of the JWS ASCII (§8). */
  id: string
}

// ── base64url ────────────────────────────────────────────────────

function b64uEncode(bytes: Uint8Array): string {
  const s = Buffer.from(bytes).toString('base64')
  // String#replaceAll with literal args: no regex (avoids a ReDoS-shaped
  // pattern) and byte-identical for base64 (padding '=' is trailing-only).
  return s.replaceAll('+', '-').replaceAll('/', '_').replaceAll('=', '')
}

function b64uDecode(s: string): Uint8Array<ArrayBuffer> {
  const pad = s.length % 4 === 0 ? '' : '='.repeat(4 - (s.length % 4))
  const b64 = s.replace(/-/g, '+').replace(/_/g, '/') + pad
  const buf = Buffer.from(b64, 'base64')
  const out = new Uint8Array(buf.length)
  out.set(buf)
  return out
}

function enc(s: string): Uint8Array<ArrayBuffer> {
  const u = new TextEncoder().encode(s)
  const out = new Uint8Array(u.length)
  out.set(u)
  return out
}

async function sha256Hex(s: string): Promise<string> {
  const digest = await subtle.digest('SHA-256', enc(s))
  return Buffer.from(digest).toString('hex')
}

// ── validation ───────────────────────────────────────────────────
// Lenient where the Python reference is lenient: enforces structural
// invariants (typ literal, required fields per operation, parent rules),
// not value shapes.

export function validatePayload(p: ReceiptPayload): void {
  for (const k of Object.keys(p)) {
    if (!KNOWN_FIELDS.has(k) && !k.startsWith('x_')) {
      throw new Error(`reserved/unknown field: ${k} (§14)`)
    }
  }
  if (p.typ !== undefined && p.typ !== JWS_TYP) {
    throw new Error(`payload typ must be ${JWS_TYP}`)
  }
  if (!VALID_OPERATIONS.has(p.operation)) {
    throw new Error(`unknown operation: ${p.operation}`)
  }
  const has = (k: keyof ReceiptPayload) =>
    typeof p[k] === 'string' && (p[k] as string).length > 0
  if (p.operation === 'guardrail_scan' && !has('guardrail_verdict')) {
    throw new Error('guardrail_scan requires guardrail_verdict (§4)')
  }
  if (p.operation === 'guardrail_scan' && !has('ruleset_hash')) {
    throw new Error('guardrail_scan requires ruleset_hash (§4)')
  }
  if (p.operation === 'model_call' && !has('model')) {
    throw new Error('model_call requires model (§4)')
  }
  if ((p.operation === 'tool_call' || p.operation === 'sanitisation') && !has('tool')) {
    throw new Error(`${p.operation} requires tool (§4)`)
  }
  if ((p.operation === 'retrieval' || p.operation === 'sanitisation') && !has('corpus')) {
    throw new Error(`${p.operation} requires corpus (§4)`)
  }
  if (p.operation === 'user_input' && p.parents.length > 0) {
    throw new Error('user_input must have no parents')
  }
  if (p.operation !== 'user_input' && p.parents.length === 0) {
    throw new Error(`${p.operation} requires at least one parent`)
  }
}

/**
 * Build the canonical payload object, injecting the constant iss/typ and
 * sorting parents+taint, exactly as Python's
 * ProvenanceReceipt.payload() does. Empty optional string fields are
 * omitted.
 */
function payloadToObject(p: ReceiptPayload): Record<string, unknown> {
  const out: Record<string, unknown> = {
    iss: ISS,
    typ: JWS_TYP,
    iat: p.iat,
    agent_id: p.agent_id,
    agent_key_id: p.agent_key_id,
    operation: p.operation,
    parents: [...p.parents].sort(byCodeUnit),
    taint: [...p.taint].sort(byCodeUnit),
  }
  const optional: (keyof ReceiptPayload)[] = [
    'input_hash', 'output_hash', 'model', 'tool', 'corpus',
    'ruleset_hash', 'guardrail_verdict',
  ]
  for (const k of optional) {
    const v = p[k]
    if (typeof v === 'string' && v.length > 0) out[k] = v
  }
  if (p.tenant !== undefined) out.tenant = p.tenant
  for (const [k, v] of Object.entries(p)) {
    if (k.startsWith('x_')) out[k] = v
  }
  return out
}

// ── emit + verify ─────────────────────────────────────────────────

export async function emit(
  payload: ReceiptPayload,
  privateKey: webcrypto.CryptoKey,
): Promise<Receipt> {
  validatePayload(payload)
  const obj = payloadToObject(payload)

  const header = {
    alg: 'EdDSA',
    typ: JWS_TYP,
    kid: payload.agent_key_id,
    crit: [...JWS_CRIT],
    'raucle/v1': 'provenance',
  }
  const headerB = canonicalEncode(header)
  const payloadB = canonicalEncode(obj)
  const signingInput = b64uEncode(headerB) + '.' + b64uEncode(payloadB)

  const sig = new Uint8Array(
    await subtle.sign('Ed25519', privateKey, enc(signingInput)),
  )
  const jws = signingInput + '.' + b64uEncode(sig)
  const id = 'sha256:' + (await sha256Hex(jws))
  return { jws, payload: obj as ReceiptPayload, id }
}

export async function verify(
  jws: string,
  publicKey: webcrypto.CryptoKey,
): Promise<Receipt> {
  const parts = jws.split('.')
  if (parts.length !== 3) throw new Error('JWS must have three segments')
  const [headerB, payloadB, sigB] = parts
  const header = JSON.parse(new TextDecoder().decode(b64uDecode(headerB)))

  if (header.alg !== 'EdDSA') throw new Error(`unsupported alg: ${header.alg}`)
  if (header.typ !== JWS_TYP) throw new Error(`unexpected typ: ${header.typ}`)
  if (!Array.isArray(header.crit) || header.crit.length !== 1 || header.crit[0] !== 'raucle/v1') {
    throw new Error("crit must be exactly ['raucle/v1']")
  }
  if (header['raucle/v1'] !== 'provenance') {
    throw new Error("header 'raucle/v1' must be 'provenance'")
  }
  const allowedHeaderKeys = new Set(['alg', 'typ', 'kid', 'crit', 'raucle/v1'])
  for (const k of Object.keys(header)) {
    if (!allowedHeaderKeys.has(k)) throw new Error(`unexpected JOSE header key: ${k}`)
  }

  const signingInput = enc(headerB + '.' + payloadB)
  const ok = await subtle.verify('Ed25519', publicKey, b64uDecode(sigB), signingInput)
  if (!ok) throw new Error('signature invalid')

  const payloadObj = JSON.parse(
    new TextDecoder().decode(b64uDecode(payloadB)),
  ) as ReceiptPayload
  validatePayload(payloadObj)

  if (header.kid !== payloadObj.agent_key_id) {
    throw new Error('header.kid != payload.agent_key_id (§3)')
  }

  const id = 'sha256:' + (await sha256Hex(jws))
  return { jws, payload: payloadObj, id }
}
